tests/integration/awslambda/test_lambda_whitebox.py

```diff
@@ -370,3 +382,57 @@ def test_python3_runtime_multiple_create_with_conflicting_module(self):
             testutil.delete_lambda_function(lambda_name2)
         finally:
             lambda_api.DO_USE_DOCKER = original_do_use_docker
+
+
+class TestFunctionStates:
+    def test_invoke_failure_when_state_pending(self, lambda_client, lambda_su_role, monkeypatch):
+        """Tests if a lambda invocation fails if state is pending"""
+        function_name = f"test-function-{short_uid()}"
+        zip_file = create_lambda_archive(load_file(TEST_LAMBDA_PYTHON_ECHO), get_content=True)
+
+        function_code_set = threading.Event()
+
+        def _do_set_function_code(*args, **kwargs):
+            result = do_set_function_code(*args, **kwargs)
+            function_code_set.wait()
+            return result
+
+        monkeypatch.setattr(
+            localstack.services.awslambda.lambda_api, "do_set_function_code", _do_set_function_code
+        )
+        try:
+            response = lambda_client.create_function(
+                FunctionName=function_name,
+                Runtime="python3.9",
+                Handler="handler.handler",
+                Role=lambda_su_role,
+                Code={"ZipFile": zip_file},
+            )
+
+            assert response["State"] == "Pending"
+
+            with pytest.raises(ClientError) as e:
+                lambda_client.invoke(FunctionName=function_name, Payload=b"{}")
+
+            assert e.value.response["ResponseMetadata"]["HTTPStatusCode"] == 409
+            assert e.match("ResourceConflictException")
+            assert e.match(
+                "The operation cannot be performed at this time. The function is currently in the following state: Pending"
+            )
+
+            # let function move to active
+            function_code_set.set()
+
+            # lambda has to get active at some point
+            def _check_lambda_state():
+                response = lambda_client.get_function(FunctionName=function_name)
+                assert response["Configuration"]["State"] == "Active"
+                return response
+
+            retry(_check_lambda_state)
+            lambda_client.invoke(FunctionName=function_name, Payload=b"{}")
+        finally:
+            try:
+                lambda_client.delete_function(FunctionName=function_name)
```

Comment:
Side note: looks like there could be an opportunity for a `create_lambda_function` fixture to take care of creating a test Lambda and automatically performing the cleanup. (not for this PR, but maybe in one of the next iterations.. 👍 ).

Reply:
We actually have such a fixture (called create_lambda_function), but the fixture automatically waits until the function is ready, which is a behavior we do not want here - which is why try: finally: is necessary in this test.